Persistent crash in Raft

I was trying to start the first of the three nodes in a local cluster (the other 2 nodes were down), and I got a panic. It looks like a Replica's mu.raftGroup is somehow nil. Perhaps it's because of the errors that precede the panic?

This is consistently reproducible with my set of 'data' directories, which I've packaged up here:

https://www.dropbox.com/s/qzw3d8eh62ujhqx/bad_data.tgz?dl=0

Below is the Cockroach log output. Let me know if you need any other info.

```
/Users/cdo/gocode/bin/cockroach start --insecure --stores=ssd=data1 --addr=127.0.0.1:26001 --pgaddr=127.0.0.1:15001 --gossip=127.0.0.1:26001,127.0.0.2:26002,127.0.0.3:26003
I0113 09:45:43.061175 31993 cli/start.go:141  build Vers: go1.5.2
I0113 09:45:43.061271 31993 cli/start.go:142  build Tag:  v0.1-alpha-359-g80581b9
I0113 09:45:43.061282 31993 cli/start.go:143  build Time: 2016/01/13 14:45:22
I0113 09:45:43.061290 31993 cli/start.go:144  build Deps: github.com/VividCortex/ewma:c34099b489e4ac33ca8d8c5f9d29d6eeaf69f2ed github.com/biogo/store:3b4c041f52c224ee4a44f5c8b150d003a40643a0 github.com/cockroachdb/c-lz4:c40aaae2fc50293eb8750b34632bc3efe813e23f github.com/cockroachdb/c-protobuf:4feb192131ea08dfbd7253a00868ad69cbb61b81 github.com/cockroachdb/c-rocksdb:b7fb7bddcb55be35eacdf67e9e2c931083ce02c4 github.com/cockroachdb/c-snappy:5c6d0932e0adaffce4bfca7bdf2ac37f79952ccf github.com/cockroachdb/cockroach:80581b9b16187c17d4728669b96fc1feb9ac092c github.com/codahale/hdrhistogram:954f16e8b9ef0e5d5189456aa4c1202758e04f17 github.com/coreos/etcd:b35ab33045197fb31d22212e9e429e6cf72ea3db github.com/elazarl/go-bindata-assetfs:57eb5e1fc594ad4b0b1dbea7b286d299e0cb43c2 github.com/gogo/protobuf:c57e439bad574c2e0877ff18d514badcfced004d github.com/google/btree:cc6329d4279e3f025a53a83c397d2339b5705c45 github.com/julienschmidt/httprouter:21439ef4d70ba4f3e2a5ed9249e7b03af4019b40 github.com/lib/pq:2d785ad24be7a038e2fc35424b993645a0025425 github.com/mattn/go-runewidth:e882a96ec18dd43fa283187b66af74497c9101c0 github.com/montanaflynn/stats:2c10aa99e7ec8c4607d4427ec7a1a60fcdfce85f github.com/olekukonko/tablewriter:48dc4474bcf3e0134e9a64222207b1e020f171e9 github.com/peterh/liner:3f1c20449d1836aa4cbe38731b96f95cdf89634d github.com/rcrowley/go-metrics:7839c01b09d2b1d7068034e5fe6e423f6ac5be22 github.com/spf13/cobra:97f831a4fd01378ad85840fed4f43824bc3dd102 github.com/spf13/pflag:7f60f83a2c81bc3c3c0d5297f61ddfa68da9d3b7 golang.org/x/crypto:d67eb63455fa4d6fca5802332d86f1f204017e00 golang.org/x/net:76365a41624aa608f888301bf60fabecaf928b5a golang.org/x/text:cf4986612c83df6c55578ba198316d1684a9a287 gopkg.in/yaml.v1:9f9df34309c04878acc86042b16630b0f696e1de
I0113 09:45:43.061370 31993 server/context.go:186  1 storage engine(s) specified
I0113 09:45:43.061392 31993 cli/start.go:176  starting cockroach node
W0113 09:45:43.061412 31993 server/server.go:102  running in insecure mode, this is strongly discouraged. See --insecure and --certs.
I0113 09:45:43.063136 31993 storage/engine/rocksdb.go:106  opening rocksdb instance at "data1"
I0113 09:45:43.063152 31993 gossip/gossip.go:916  starting client to 127.0.0.2:26002
W0113 09:45:43.063480 31993 rpc/client.go:317  dial tcp 127.0.0.2:26002: getsockopt: connection refused
I0113 09:45:43.093544 31993 server/node.go:293  initialized store store=1:1 ([ssd]=data1): {Capacity:249779191808 Available:220275060736 RangeCount:0}
I0113 09:45:43.093736 31993 server/node.go:220  node ID 1 initialized
I0113 09:45:43.093870 31993 gossip/gossip.go:218  setting node descriptor node_id:1 address:<network_field:"tcp" address_field:"127.0.0.1:26001" > attrs:<> 
W0113 09:45:43.094056 31993 storage/replica.go:1576  could not acquire lease for range gossip: range 1 was not found
W0113 09:45:43.094227 31993 storage/replica.go:1576  could not acquire lease for range gossip: range 1 was not found
W0113 09:45:43.094595 31993 storage/store.go:639  error gossiping system config: range 1 was not found
W0113 09:45:43.094613 31993 storage/store.go:620  error gossiping first range data: range 1 was not found
I0113 09:45:43.101676 31993 gossip/gossip.go:267  read 3 gossip host(s) for bootstrapping from persistent storage
I0113 09:45:43.101704 31993 server/node.go:389  connecting to gossip network to verify cluster ID...
I0113 09:45:43.101721 31993 server/node.go:406  node connected via gossip and verified as part of cluster "3ce7a37d-7215-41a2-9411-318829882572"
I0113 09:45:43.101772 31993 server/node.go:258  Started node with [[ssd]=data1] engine(s) and attributes []
I0113 09:45:43.101828 31993 server/server.go:262  starting http server at 127.0.0.1:26001
I0113 09:45:43.101942 31993 sql/pgwire/server.go:76  starting postgres server at 127.0.0.1:15001
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x0 pc=0x46f7634]

goroutine 51 [running]:
github.com/coreos/etcd/raft.(*RawNode).Tick(0x0)
	/Users/cdo/gocode/src/github.com/coreos/etcd/raft/rawnode.go:115 +0x14
github.com/cockroachdb/cockroach/storage.(*Store).processRaft.func1()
	/Users/cdo/gocode/src/github.com/cockroachdb/cockroach/storage/store.go:1604 +0x790
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc820214a80, 0xc8203717e0)
	/Users/cdo/gocode/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:89 +0x52
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
	/Users/cdo/gocode/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:90 +0x62
```

--- a/storage/client_raft_test.go
+++ b/storage/client_raft_test.go
@@ -1411,3 +1411,67 @@ func TestLeaderRemoveSelf(t *testing.T) {
 		t.Fatalf("expect get RangeNotFoundError, actual get %v ", pErr)
 	}
 }
+
+// TestRemoveRangeWithoutGC ensures that we do not panic when a
+// replica has been removed but not yet GC'd (and therefore
+// does not have an active raft group).
+func TestRemoveRangeWithoutGC(t *testing.T) {
+	defer leaktest.AfterTest(t)
+
+	mtc := startMultiTestContext(t, 2)
+	defer mtc.Stop()
+	// Disable the GC queue and move the range from store 0 to 1.
+	mtc.stores[0].DisableReplicaGCQueue(true)
+	const rangeID roachpb.RangeID = 1
+	mtc.replicateRange(rangeID, 1)
+	mtc.unreplicateRange(rangeID, 0)
+
+	// Wait for store 0 to process the removal.
+	util.SucceedsWithin(t, time.Second, func() error {
+		rep, err := mtc.stores[0].GetReplica(rangeID)
+		if err != nil {
+			return err
+		}
+		desc := rep.Desc()
+		if len(desc.Replicas) != 1 {
+			return util.Errorf("range has %d replicas", len(desc.Replicas))
+		}
+		return nil
+	})
+
+	// The replica's data is still on disk even though the Replica
+	// object is removed.
+	var desc roachpb.RangeDescriptor
+	descKey := keys.RangeDescriptorKey(roachpb.RKeyMin)
+	if ok, err := engine.MVCCGetProto(mtc.stores[0].Engine(), descKey,
+		mtc.stores[0].Clock().Now(), true, nil, &desc); err != nil {
+		t.Fatal(err)
+	} else if !ok {
+		t.Fatal("expected range descriptor to be present")
+	}
+
+	// Stop and restart the store to reset the replica's raftGroup
+	// pointer to nil. As long as the store has not been restarted it
+	// can continue to use its last known replica ID.
+	mtc.stopStore(0)
+	mtc.restartStore(0)
+	// Turn off the GC queue to ensure that the replica is deleted at
+	// startup instead of by the scanner. This is not 100% guaranteed
+	// since the scanner could have already run at this point, but it
+	// should be enough to prevent us from accidentally relying on the
+	// scanner.
+	mtc.stores[0].DisableReplicaGCQueue(true)
+
+	// The Replica object is not recreated.
+	if _, err := mtc.stores[0].GetReplica(rangeID); err == nil {
+		t.Fatalf("expected replica to be missing")
+	}
+
+	// And the data is no longer on disk.
+	if ok, err := engine.MVCCGetProto(mtc.stores[0].Engine(), descKey,
+		mtc.stores[0].Clock().Now(), true, nil, &desc); err != nil {
+		t.Fatal(err)
+	} else if ok {
+		t.Fatal("expected range descriptor to be absent")
+	}
+}

--- a/storage/replica.go
+++ b/storage/replica.go
@@ -179,15 +179,18 @@ type Replica struct {
 
 var _ client.Sender = &Replica{}
 
-// NewReplica initializes the replica using the given metadata.
-func NewReplica(desc *roachpb.RangeDescriptor, store *Store) (*Replica, error) {
+// NewReplica initializes the replica using the given metadata. If the
+// replica is initialized (i.e. desc contains more than a RangeID),
+// replicaID should be 0 and the replicaID will be discovered from the
+// descriptor.
+func NewReplica(desc *roachpb.RangeDescriptor, store *Store, replicaID roachpb.ReplicaID) (*Replica, error) {
 	r := &Replica{
 		store:    store,
 		sequence: NewSequenceCache(desc.RangeID),
 		RangeID:  desc.RangeID,
 	}
 
-	if err := r.newReplicaInner(desc, store.Clock()); err != nil {
+	if err := r.newReplicaInner(desc, store.Clock(), replicaID); err != nil {
 		return nil, err
 	}
 
@@ -201,7 +204,7 @@ func NewReplica(desc *roachpb.RangeDescriptor, store *Store) (*Replica, error) {
 	return r, nil
 }
 
-func (r *Replica) newReplicaInner(desc *roachpb.RangeDescriptor, clock *hlc.Clock) error {
+func (r *Replica) newReplicaInner(desc *roachpb.RangeDescriptor, clock *hlc.Clock, replicaID roachpb.ReplicaID) error {
 	r.mu.Lock()
 	defer r.mu.Unlock()
 
@@ -227,14 +230,18 @@ func (r *Replica) newReplicaInner(desc *roachpb.RangeDescriptor, clock *hlc.Cloc
 		return err
 	}
 
-	_, repDesc := desc.FindReplica(r.store.StoreID())
-	if repDesc != nil {
-		if err := r.setReplicaIDLocked(repDesc.ReplicaID); err != nil {
-			return err
-		}
+	if r.isInitializedLocked() && replicaID != 0 {
+		return util.Errorf("replicaID must be 0 when creating an initialized replica")
 	}
-
-	return nil
+	if replicaID == 0 {
+		_, repDesc := desc.FindReplica(r.store.StoreID())
+		if repDesc == nil {
+			return util.Errorf("cannot recreate replica that is not a member of its range (StoreID %s not found in %s)",
+				r.store.StoreID(), desc)
+		}
+		replicaID = repDesc.ReplicaID
+	}
+	return r.setReplicaIDLocked(replicaID)
 }
 
 // String returns a string representation of the range.
@@ -250,25 +257,7 @@ func (r *Replica) Destroy(origDesc roachpb.RangeDescriptor) error {
 		return util.Errorf("cannot destroy replica %s; replica ID has changed (%s >= %s)",
 			r, rd.ReplicaID, origDesc.NextReplicaID)
 	}
-	iter := newReplicaDataIterator(desc, r.store.Engine())
-	defer iter.Close()
-	batch := r.store.Engine().NewBatch()
-	defer batch.Close()
-	for ; iter.Valid(); iter.Next() {
-		_ = batch.Clear(iter.Key())
-	}
-
-	// Save a tombstone. The range cannot be re-replicated onto this
-	// node without having a replica ID of at least desc.NextReplicaID.
-	tombstoneKey := keys.RaftTombstoneKey(desc.RangeID)
-	tombstone := &roachpb.RaftTombstone{
-		NextReplicaID: desc.NextReplicaID,
-	}
-	if err := engine.MVCCPutProto(batch, nil, tombstoneKey, roachpb.ZeroTimestamp, nil, tombstone); err != nil {
-		return err
-	}
-
-	return batch.Commit()
+	return r.store.destroyReplicaData(desc)
 }
 
 func (r *Replica) setReplicaID(replicaID roachpb.ReplicaID) error {

--- a/storage/replica_command.go
+++ b/storage/replica_command.go
@@ -1397,7 +1397,7 @@ func (r *Replica) splitTrigger(batch engine.Engine, split *roachpb.SplitTrigger)
 	// Add the new split replica to the store. This step atomically
 	// updates the EndKey of the updated replica and also adds the
 	// new replica to the store's replica map.
-	newRng, err := NewReplica(&split.NewDesc, r.store)
+	newRng, err := NewReplica(&split.NewDesc, r.store, 0)
 	if err != nil {
 		return err
 	}

--- a/storage/replica_test.go
+++ b/storage/replica_test.go
@@ -164,7 +164,7 @@ func (tc *testContext) Start(t testing.TB) {
 
 	if realRange {
 		if tc.bootstrapMode == bootstrapRangeOnly {
-			rng, err := NewReplica(testRangeDescriptor(), tc.store)
+			rng, err := NewReplica(testRangeDescriptor(), tc.store, 0)
 			if err != nil {
 				t.Fatal(err)
 			}
@@ -248,6 +248,12 @@ func TestRangeContains(t *testing.T) {
 		RangeID:  1,
 		StartKey: roachpb.RKey("a"),
 		EndKey:   roachpb.RKey("b"),
+		Replicas: []roachpb.ReplicaDescriptor{{
+			// This test is a little weird since it doesn't call Store.Start
+			// which would assign IDs.
+			StoreID:   0,
+			ReplicaID: 1,
+		}},
 	}
 
 	stopper := stop.NewStopper()
@@ -259,7 +265,7 @@ func TestRangeContains(t *testing.T) {
 	ctx.Transport = NewLocalRPCTransport(stopper)
 	defer ctx.Transport.Close()
 	store := NewStore(ctx, e, &roachpb.NodeDescriptor{NodeID: 1})
-	r, err := NewReplica(desc, store)
+	r, err := NewReplica(desc, store, 0)
 	if err != nil {
 		t.Fatal(err)
 	}
@@ -486,7 +492,7 @@ func TestRangeLeaderLease(t *testing.T) {
 	}
 
 	// Verify that command returns NotLeaderError when lease is rejected.
-	rng, err := NewReplica(testRangeDescriptor(), tc.store)
+	rng, err := NewReplica(testRangeDescriptor(), tc.store, 0)
 	if err != nil {
 		t.Fatal(err)
 	}
@@ -3418,7 +3424,7 @@ func TestRequestLeaderEncounterGroupDeleteError(t *testing.T) {
 	proposeRaftCommandFn := func(cmdIDKey, roachpb.RaftCommand) error {
 		return &roachpb.RaftGroupDeletedError{}
 	}
-	rng, err := NewReplica(testRangeDescriptor(), tc.store)
+	rng, err := NewReplica(testRangeDescriptor(), tc.store, 0)
 	rng.proposeRaftCommandFn = proposeRaftCommandFn
 	if err != nil {
 		t.Fatal(err)

--- a/storage/store.go
+++ b/storage/store.go
@@ -531,7 +531,15 @@ func (s *Store) Start(stopper *stop.Stopper) error {
 			if err := kv.Value.GetProto(&desc); err != nil {
 				return false, err
 			}
-			rng, err := NewReplica(&desc, s)
+
+			if _, repDesc := desc.FindReplica(s.StoreID()); repDesc == nil {
+				// We are no longer a member of the range, but we didn't GC
+				// the replica before shutting down. Destroy the replica now
+				// to avoid creating a new replica without a valid replica ID
+				// (which is necessary to have a non-nil raft group)
+				return false, s.destroyReplicaData(&desc)
+			}
+			rng, err := NewReplica(&desc, s, 0)
 			if err != nil {
 				return false, err
 			}
@@ -539,6 +547,7 @@ func (s *Store) Start(stopper *stop.Stopper) error {
 				return false, err
 			}
 			s.feed.registerRange(rng, true /* scan */)
+			// TODO(bdarnell): lazily create raft groups to make the following comment true again.
 			// Note that we do not create raft groups at this time; they will be created
 			// on-demand the first time they are needed. This helps reduce the amount of
 			// election-related traffic in a cold start.
@@ -1140,6 +1149,30 @@ func (s *Store) removeReplicaImpl(rep *Replica, origDesc roachpb.RangeDescriptor
 	return nil
 }
 
+// destroyReplicaData deletes all data associated with a replica, leaving a tombstone.
+// If a Replica object exists, use Replica.Destroy instead of this method.
+func (s *Store) destroyReplicaData(desc *roachpb.RangeDescriptor) error {
+	iter := newReplicaDataIterator(desc, s.Engine())
+	defer iter.Close()
+	batch := s.Engine().NewBatch()
+	defer batch.Close()
+	for ; iter.Valid(); iter.Next() {
+		_ = batch.Clear(iter.Key())
+	}
+
+	// Save a tombstone. The range cannot be re-replicated onto this
+	// node without having a replica ID of at least desc.NextReplicaID.
+	tombstoneKey := keys.RaftTombstoneKey(desc.RangeID)
+	tombstone := &roachpb.RaftTombstone{
+		NextReplicaID: desc.NextReplicaID,
+	}
+	if err := engine.MVCCPutProto(batch, nil, tombstoneKey, roachpb.ZeroTimestamp, nil, tombstone); err != nil {
+		return err
+	}
+
+	return batch.Commit()
+}
+
 // processRangeDescriptorUpdate is called whenever a range's
 // descriptor is updated.
 func (s *Store) processRangeDescriptorUpdate(rng *Replica) error {
@@ -1665,11 +1698,8 @@ func (s *Store) getOrCreateReplicaLocked(groupID roachpb.RangeID, replicaID roac
 		RangeID: groupID,
 		// TODO(bdarnell): other fields are unknown; need to populate them from
 		// snapshot.
-	}, s)
+	}, s, replicaID)
 	if err != nil {
-		return nil, err
-	}
-	if err := r.setReplicaID(replicaID); err != nil {
 		return nil, err
 	}
 	// Add the range to range map, but not rangesByKey since

--- a/storage/store_test.go
+++ b/storage/store_test.go
@@ -260,8 +260,14 @@ func createRange(s *Store, rangeID roachpb.RangeID, start, end roachpb.RKey) *Re
 		RangeID:  rangeID,
 		StartKey: start,
 		EndKey:   end,
+		Replicas: []roachpb.ReplicaDescriptor{{
+			NodeID:    1,
+			StoreID:   1,
+			ReplicaID: 1,
+		}},
+		NextReplicaID: 2,
 	}
-	r, err := NewReplica(desc, s)
+	r, err := NewReplica(desc, s, 0)
 	if err != nil {
 		log.Fatal(err)
 	}
@@ -711,7 +717,7 @@ func splitTestRange(store *Store, key, splitKey roachpb.RKey, t *testing.T) *Rep
 	if err != nil {
 		t.Fatal(err)
 	}
-	newRng, err := NewReplica(desc, store)
+	newRng, err := NewReplica(desc, store, 0)
 	if err != nil {
 		t.Fatal(err)
 	}

--- a/storage/stores_test.go
+++ b/storage/stores_test.go
@@ -155,7 +155,7 @@ func TestStoresLookupReplica(t *testing.T) {
 			EndKey:   rng.end,
 			Replicas: []roachpb.ReplicaDescriptor{{StoreID: rng.storeID}},
 		}
-		newRng, err := NewReplica(d[i], s[i])
+		newRng, err := NewReplica(d[i], s[i], 0)
 		if err != nil {
 			t.Fatal(err)
 		}

